### PR TITLE
Added pm_live_screens for live screen reloading!

### PR DESCRIPTION
**Pull request by jamonholmgren**
Live code reloading in screens works great.

gif showing how

### CODE REVIEW

**Comment by markrickert** [comment]
Holee crap!

**Comment by GantMan** [comment]
Burn the witch!

**Comment by squidpunch** [comment]
nice job @jamonholmgren

File: docs/Guides/Live Screen Reloading.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    load_data
+  end
+
+  def load_data
+    @some_live_data = { jamon: "Holmgren" }
+  end
+
+  def on_live_reload
+    # teardown all views and data
+    self.view.subviews.each(&:removeFromSuperview)
```
Comment (squidpunch) [inline docs/Guides/Live Screen Reloading.md]: should we do this in the base by default?  So if I have no extra setup on my screen, i turn on live screen loading and then the views are removed and re-added properly for me without needing to update x screens in my existing project with this new method?  Obviously consumers can override the method as they see fit, to expand if they need it to do more (like removing `@some_live_data` as you shown here?

**Comment by ryanlntn** [comment]
:+1: Awesome.

File: docs/Guides/Live Screen Reloading.md (hunk 1 above)
Comment (ryanlntn) [inline docs/Guides/Live Screen Reloading.md]: :+1:

Comment (jamonholmgren) [inline docs/Guides/Live Screen Reloading.md]: I think you're right, @squidpunch. Do the subviews remove and then the `on_load`.

File: lib/ProMotion/repl/repl.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            eval(new_code)
+            callback.call live_file, new_code
+            live_files[live_file] = File.mtime(live_file)
+          end
+        end
+      end
+
+      "Watching #{path_query}."
+      live_reload_timer
+    end
```
Comment (jamonholmgren) [inline lib/ProMotion/repl/repl.rb]: @twerth This method is more general-purpose and could be used for RMQ's live reload as well. We should probably extract it into its own gem.

You would use it like this:

```ruby
@my_timer.invalidate if @my_timer
@my_timer = pm_live_watch("app/subfolder") do |reloaded_file, reloaded_code|
  rmq.screen.children.and_self.reapply_styles
end
```

Alternatively, you could do some class detection with the new code (like I'm doing above on line 10) and more intelligently reload things.

There's a lot of potential here.

**Comment by markrickert** [comment]
Any more progress on this or are you focusing on Android stuff at the moment?

**Comment by jamonholmgren** [comment]
I'm focusing on Android, but there's really only one thing left to do, and that's add a default handler for `PM::Screen` and `PM::TableScreen`.

```ruby
# screen
def on_live_reload
  self.view.subviews.each(&:removeFromSuperview)
  on_load
end
```

```ruby
# table screen
def on_live_reload
  update_table_data
end
```

**Comment by jamonholmgren** [comment]
After that we should be able to merge and then hook it into `live` on the RedPotion side.

**Comment by squidpunch** [comment]
I just pushed a commit to add the base handlers, and tweak the documentation a bit to mention its possible to use the base handler if you aren't getting fancy.

Not really something we can write tests for, but I did spot check it on a table and regular screen without tear down handlers and worked!

**Comment by markrickert** [comment]
Delete branch? :+1: / :-1:

**Comment by jamonholmgren** [comment]
Yes, I thought initially I wouldn't but I don't want people using this branch anymore.

**Comment by markrickert** [comment]
Yup. Lots of new fixes and stuff in master.

**Pull request merged**